src/plugins/mer/mersettings.cpp

```diff
@@ -123,6 +125,119 @@ QMap<QString, MerEmulatorDeviceModel> MerSettings::deviceModelsRead(const Utils:
     return result;
 }
 
+FileName MerSettings::deviceModelsFileName()
+{
+    const QFileInfo settingsLocation(PluginManager::settings()->fileName());
+    return FileName::fromString(settingsLocation.absolutePath() + QLatin1String(MER_DEVICE_MODELS_FILENAME));
+}
+
+QString MerSettings::currentDeviceModel(Core::Id emulatorDeviceId)
+{
+    Q_ASSERT(s_instance);
+
+    return s_instance->m_currentDeviceModels.value(emulatorDeviceId);
+}
+
+void MerSettings::setCurrentDeviceModel(Core::Id emulatorDeviceId, const QString &name)
+{
+    Q_ASSERT(s_instance);
+
+    s_instance->m_currentDeviceModels.insert(emulatorDeviceId, name);
+}
+
+bool MerSettings::updateDeviceModel(const MerEmulatorDeviceModel &model)
+{
+    Q_ASSERT(s_instance);
+    QTC_ASSERT(!model.isSdkProvided(), return false);
+
+    s_instance->m_deviceModels.insert(model.name(), model);
+    emit s_instance->deviceModelUpdated(model.name());
+
+    return true;
+}
+
+bool MerSettings::addDeviceModel(const QString &name)
+{
+    Q_ASSERT(s_instance);
+    QTC_ASSERT(!s_instance->m_deviceModels.contains(name), return false);
+
+    MerEmulatorDeviceModel emulatorProfile(name, QSize(500, 900), QSize(50, 90));
+    s_instance->m_deviceModels.insert(name, emulatorProfile);
+    emit s_instance->deviceModelAdded(name);
+
+    return true;
+}
+
+bool MerSettings::removeDeviceModel(const QString &name)
+{
+    Q_ASSERT(s_instance);
+    QTC_ASSERT(s_instance->m_deviceModels.contains(name), return false);
+    QTC_ASSERT(!s_instance->m_deviceModels.value(name).isSdkProvided(), return false);
+
+    s_instance->m_deviceModels.remove(name);
+    emit s_instance->deviceModelRemoved(name);
+
+    return true;
+}
+
+bool MerSettings::isDeviceModelStored(const QString &name)
+{
+    Q_ASSERT(s_instance);
+
+    if (deviceModelsRead(globalDeviceModelsFileName()).contains(name))
+        return true;
+
+    const QMap<QString, MerEmulatorDeviceModel> models = deviceModelsRead(deviceModelsFileName());
+    return models.contains(name)
+            && models.value(name) == s_instance->m_deviceModels.value(name);
+}
+
+void MerSettings::deviceModelsReset()
+{
+    Q_ASSERT(s_instance);
+
+    s_instance->m_deviceModels = deviceModelsRead(globalDeviceModelsFileName());
+    const QMap<QString, MerEmulatorDeviceModel> userDeviceModels = deviceModelsRead(deviceModelsFileName());
+
+    QTC_CHECK(s_instance->m_deviceModels.keys().toSet()
+              .intersects(userDeviceModels.keys().toSet()) == false);
+
+    s_instance->m_deviceModels.unite(userDeviceModels);
```

Comment:
What would happen if the user adds a device model and with the next release a sdk-provided model will be added using the same name as the user used?

Reply:
Then user defined device model will change name by adding index number, notifying user by MessageManager::write.